md-proto/WebAssembly.md

```diff
@@ -0,0 +1,1884 @@
+
+### Unknown Sections
+
+Modules may also contain unknown sections, with names other than those
+corresponding to known sections above. They have no semantic effect.
+
+### Function Bodies
+
+Function bodies contain an [array] of [types], which are declarations for
+locals, and a sequence of instructions.
+
+##### Positions within a function body
+
+A *position* within a function refers to an element of the sequence, or to an
+additional *return position* which is sequenced after the end of the function body.
+
+> In the binary encoding, positions are represented as byte offsets; in the text
+format, positions are represented with a special syntax.
+
+
+## Validation
+
+### Module Validation
+
+TODO: describe module validation
+
+### Function Validation
+
+For the duration of the validation of a function body, several data structures
+are created:
+ - A *control-flow stack*, with entries containing
+    - A [control-flow type](#control-flow-type).
+    - A *depth* value.
+ - A *[type] stack*.
+ - A *current position*.
+
+During validation, if either stack is empty when an element is to be popped from
+it, validation fails.
+
+#### Control Flow Type
+
+A *control-flow type* is either a [type](#types), `any`, which satisfies any
+type check, or `void`, which satisfies no check.
+
+> `any` is effectively serving as a kind of bottom type, and `void` is
+effectively serving as a kind of top type. Note that in some type theory
+contexts, it's common to use these names for exactly opposite purposes.
+
+#### Control Flow Type Merge
+
+To merge two [control-flow types]:
+ - If either control-flow type is `any`, the result is the other control flow
+   type.
+ - Otherwise, if the control-flow types are the same, the result is that control
+   flow type.
+ - Otherwise, validation fails with a type error.
+
+#### Function Validation Initialization
+
+The current position starts at the first position. The type stack begins empty.
+The control-flow stack starts with one entry. This entry's control-flow type is
+the first return type of the function, if it has any, or `void` otherwise. Its
+block depth is zero.
+
+#### Function Body Validation
+
+If the current position is the return position, function return validation is
+prompted and validation of the function is thereafter complete.
+
+Otherwise, the instruction at the current position is remembered, the current
+position is incremented to point to the instruction following it, or if there
+are no instructions following it, to the return position. Then the remembered
+instruction is validated as follows.
+
+If the instruction has a **Validation** clause, that clause describes its
+validation. Otherwise, the following generic validation is performed:
+
+For each operand [type] in the instruction's signature, a type is popped from
+the type stack and checked to be the same as the corresponding operand type.
+Each return type of the instruction's signature is then pushed onto the type
+stack.
+
```

Comment:
This might be ok, but this is proposing to push *no* values on the type stack if the operator returns no values, and currently a `void` type element is pushed for an operator returning no values (even for `drop` which does no change the size of the type stack).

`drop` did not change the size of the type stack for good reason, so that `i32_const 1, i32_const 2, drop, i32.eqz` is not valid - it's not really popping rather writing over the top of the stack with the `void` type. If we change the semantics to push each value to separate stack elements then I am not sure if the code would have such a simple text format with expressions. (Assuming `void` is the type of an operator returning no values)

For example, how would the second value in `call $fn_returning_i32_i64` even be invalidated (or `drop`ped), as `drop` just writes over the top-of-stack type.

Another other path to take would be to embrace this proposal and not push a type on the stack for an operator returning no values, and this would eliminate the `void` type. The `drop` operator would then pop the stack.

There are different ways this could go but this description does not appear to present a clear option, not a clear stack machine if that was the intention.

If we are going to keep the `void` type, then `drop` could be extended to invalidate any entry up the stack, not just the top.